torch/csrc/autograd/engine.cpp

```diff
@@ -858,16 +869,18 @@ auto Engine::start_threads() -> void {
 
   // One for CPU, plus one for every GPU device (but colocate GPUs of different
   // types)
-  int num_threads = num_devices + 1;
-  ready_queues_ = std::vector<std::shared_ptr<ReadyQueue>>(num_threads);
+  ++num_devices;
+  ready_queues_ = std::vector<std::shared_ptr<ReadyQueue>>(num_devices);
   for (auto& queue : ready_queues_)
     queue.reset(new ReadyQueue());
 
   thread_pool_shared_ = std::make_shared<ThreadPoolShared>();
 
-  for (int i = 0; i < num_threads; ++i) {
-    std::thread t(&Engine::thread_init, this, i - 1);
-    t.detach();
+  for (int i = 0; i < num_devices; ++i) {
+    for (size_t j = 0; j < num_threads_per_device_; ++j) {
```

Comment:
I kind of wonder if you actually want a uniform number of threads per cuda device. My understanding from Caffe2's design is that some multithreading over streams on a single gpu device is helpful, but only up to a point. In any case, I'd expect the performance characteristics for CUDA to be different from cPU.